### ISSUE TITLE
[@typescript-eslint/no-unsafe-return] Incorrectly throwing when using TSX and returning JSX.Element objects

### ISSUE THREAD

**Issue by vladfrangu** [typescript-eslint/typescript-eslint#2210]
**Repro**

```JSON
{
	"env": {
		"node": true
	},
	"extends": [
		"plugin:vue/recommended",
		"bamboo"
	],
	"rules": {
		"sort-imports": 0,
		"vue/script-indent": ["error", "tab"],
		"no-undef": 0,
		"no-unused-vars": 0,
		"strict": 0,
		"vue/html-indent": ["warn", "tab"],
		"vue/max-attributes-per-line": ["error", {
			"singleline": 3,
			"multiline": {
				"max": 3,
				"allowFirstLine": true
			}
		}],
		"@typescript-eslint/no-unsafe-assignment": 0,
		"@typescript-eslint/naming-convention": 0
	},
	"parser": "vue-eslint-parser",
	"parserOptions": {
		"parser": "@typescript-eslint/parser",
		"ecmaFeatures": {
			"jsx": true,
			"modules": true
		},
		"sourceType": "module",
		"extraFileExtensions": [
			".vue",
			".tsx"
		]
	}
}
```

```TSX
import { Component } from 'nuxt-property-decorator';
import * as tsx from 'vue-tsx-support';
import { VSnackbar, VIcon } from 'vuetify/lib';

import { SnackbarStore } from '@/lib/stores/SnackbarStore';

@Component
export class Snackbar extends tsx.Component<never> {

	public render(): JSX.Element {
		return <VSnackbar
			// v-model?
			value={SnackbarStore.shown}
			top
			absolute
			color={SnackbarStore.color}
			timeout={SnackbarStore.timeout}
		>
			<VIcon left>
				{SnackbarStore.icon}
			</VIcon>
			<span domPropsInnerHTML={SnackbarStore.message} />
		</VSnackbar>;
	}

}
```

**Expected Result** typescript-eslint infers the return type as an JSX.Element (like TS does for the render return function)

**Actual Result** typescript-eslint reports that I'm returning an unsafe any value.

**Additional Info** This error is present on **any** JSX.Element, not just custom ones. I've also added the versions for the respective rule plugins at the bottom of the versions table.

**Versions**

| package                            | version  |
| ---------------------------------- | -------- |
| `@typescript-eslint/eslint-plugin` | `3.2.0`  |
| `@typescript-eslint/parser`        | `3.2.0`  |
| `TypeScript`                       | `3.9.5`  |
| `ESLint`                           | `7.2.0`  |
| `node`                             | `14.4.0` |
| `npm`                              | `6.14.3` |
| `eslint-config-bamboo`             | `5.0.0`  |
| `eslint-plugin-vue`                | `6.2.2`  |

**Comment by bradzacher** [comment]
Could you please provide a repro repo? There's a lot of dependencies here.

**Comment by vladfrangu** [comment]
Come to think of it, that example is very much too complex; a simple reproduction step can be that eslint.json file and a mere
```tsx
export const Snackbar = () => <div>This is a div!</div>;
```

**Comment by vladfrangu** [comment]
@bradzacher I've created a sample repository that replicates this issue! Please ignore the ts error (which..doesn't happen locally but happens here... I love differences between projects). Repository is available at https://github.com/vladfrangu/typescript-eslint-2210

**Comment by bradzacher** [comment]
So digging into it - this is happening due to  your config.
The reason that TS is resolving the JSX to any is that it can't resolve the type of the JSX things. When it can't resolve the type, it emits a special error any type.

Your [tsconfig.eslint.json](https://github.com/vladfrangu/typescript-eslint-2210/blob/master/tsconfig.eslint.json) sets a new set of `include`s. This _overwrites_ the includes defined in your base config.

If make this change, it all works as expected
```diff
	"include": [
+		"node_modules/vue-tsx-support/enable-check.d.ts",
+		"node_modules/vue-tsx-support/options/enable-vue-router.d.ts",
+		"node_modules/vue-tsx-support/options/allow-unknown-props.d.ts",
		"./src/**/*.ts",
		"./src/**/*.tsx",
		"./src/**/*.vue",
		"./nuxt.config.ts"
	]
```
